integration-tests/src/test_loop/env.rs

```diff
@@ -67,6 +72,58 @@ impl TestLoopEnv {
     }
 }
 
+/// Stores all chunks ever observed on chain.
+///
+/// Needed to intercept network messages storing chunk hash only, while
+/// interception requires more detailed information like shard id.
+#[derive(Default)]
+pub struct TestLoopChunksStorage {
+    /// Mapping from chunk hashes to headers.
+    storage: HashMap<ChunkHash, ShardChunkHeader>,
+    /// Minimal chunk height ever observed.
+    pub min_chunk_height: Option<BlockHeight>,
+}
+
+impl TestLoopChunksStorage {
+    pub fn insert(&mut self, chunk_header: ShardChunkHeader) {
+        let chunk_height = chunk_header.height_created();
+        self.min_chunk_height = Some(
+            self.min_chunk_height
+                .map_or(chunk_height, |current_height| current_height.min(chunk_height)),
+        );
+        self.storage.insert(chunk_header.chunk_hash(), chunk_header);
+    }
+
+    pub fn get(&self, chunk_hash: &ChunkHash) -> Option<&ShardChunkHeader> {
+        self.storage.get(chunk_hash)
+    }
+}
+
+/// Custom implementation of `Sender` for messages from `Client` to
+/// `ShardsManagerActor` that allows to intercept all messages indicating
+/// any chunk production and storing all chunks.
+pub struct ClientToShardsManagerSender {
+    pub sender: Arc<LateBoundSender<TestLoopSender<ShardsManagerActor>>>,
+    /// Storage of chunks shared between all test loop nodes.
+    pub chunks_storage: Arc<Mutex<TestLoopChunksStorage>>,
+}
+
+impl CanSend<ShardsManagerRequestFromClient> for ClientToShardsManagerSender {
+    fn send(&self, message: ShardsManagerRequestFromClient) {
+        // `DistributeEncodedChunk` indicates that a certain chunk was produced.
+        if let ShardsManagerRequestFromClient::DistributeEncodedChunk { partial_chunk, .. } =
+            &message
+        {
+            if let PartialEncodedChunk::V2(chunk) = partial_chunk {
```

Comment:
why do you store chunks at a certain version instead of any chunks seen?

Reply:
Good catch. I wanted to tell by this that V1 is super old and can't occur in tests anyway, but here we also have cloned_header method which I can use.